File: Problems/155. Min Stack/155. Min Stack.java
import java.util.EmptyStackException;
import java.util.Stack;

class MinStack {

    private Stack<Integer> S = new Stack<>();
    private Stack<Integer> minS = new Stack<>();

    public MinStack() { }
    
    public void push(int val) {
        if (minS.size() == 0) minS.push(val);
        else if (val <= getMin()) minS.push(val);
        S.push(val);
    }
    
    public void pop() {
        if (S.size() == 0) throw new EmptyStackException();
        if (S.peek() == getMin()) minS.pop();
        S.pop();
    }
    
    public int top() {
        if (S.size() == 0) throw new EmptyStackException();
        return S.peek();
    }
    
    public int getMin() {
        if (S.size() == 0) throw new EmptyStackException();
        return minS.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: Problems/155. Min Stack/155. Min Stack.py
from collections import deque;
class MinStack:

    def __init__(self):
        self.mainS = deque([]);
        self.minS = deque([]);

    def push(self, val: int) -> None:
        if (not self.mainS or val<=self.getMin()): self.minS.append(val);
        self.mainS.append(val);

    def pop(self) -> None:
        if (self.top() == self.getMin()): self.minS.pop();
        self.mainS.pop();

    def top(self) -> int:
        return self.mainS[-1];

    def getMin(self) -> int:
        return self.minS[-1];


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()